cmd/cdi-importer/importer.go

```diff
@@ -121,51 +106,90 @@ func main() {
 		os.Exit(1)
 	}
 	if source == controller.SourceNone {
-		if contentType == string(cdiv1.DataVolumeKubeVirt) {
-			createBlankImage(imageSize, availableDestSpace, preallocation, volumeMode, filesystemOverhead)
-			preallocationApplied = preallocation
-		} else {
-			errorEmptyDiskWithContentTypeArchive()
+		err := handleEmptyImage(contentType, imageSize, availableDestSpace, preallocation, volumeMode, filesystemOverhead)
+		if err != nil {
+			klog.Errorf("%+v", err)
+			os.Exit(1)
+		}
+
+	} else {
+		exitCode := handleImport(source, contentType, volumeMode, imageSize, filesystemOverhead, preallocation)
+		if exitCode != 0 {
+			os.Exit(exitCode)
 		}
+	}
+}
+
+func handleEmptyImage(contentType string, imageSize string, availableDestSpace int64, preallocation bool, volumeMode v1.PersistentVolumeMode, filesystemOverhead float64) error {
+	var preallocationApplied bool
+
+	if contentType == string(cdiv1.DataVolumeKubeVirt) {
+		createBlankImage(imageSize, availableDestSpace, preallocation, volumeMode, filesystemOverhead)
+		preallocationApplied = preallocation
 	} else {
-		waitForReadyFile()
-		klog.V(1).Infoln("begin import process")
+		errorEmptyDiskWithContentTypeArchive()
+	}
+
+	err := importCompleteTerminationMessage(preallocationApplied)
+	return err
+}
+
+func handleImport(
+	source string,
+	contentType string,
+	volumeMode v1.PersistentVolumeMode,
+	imageSize string,
+	filesystemOverhead float64,
+	preallocation bool) int {
+	klog.V(1).Infoln("begin import process")
 
-		ds = newDataSource(source, contentType, volumeMode)
-		defer ds.Close()
+	ds := newDataSource(source, contentType, volumeMode)
+	defer func() {
+		ds.Close()
+	}()
 
-		processor := newDataProcessor(contentType, volumeMode, ds, imageSize, filesystemOverhead, preallocation)
-		err = processor.ProcessData()
+	processor := newDataProcessor(contentType, volumeMode, ds, imageSize, filesystemOverhead, preallocation)
+	waitForReadyFile()
+	err := processor.ProcessData()
 
+	if err != nil {
+		klog.Errorf("%+v", err)
+		if err == importer.ErrRequiresScratchSpace {
+			return common.ScratchSpaceNeededExitCode
+		}
+		err = util.WriteTerminationMessage(fmt.Sprintf("Unable to process data: %+v", err))
 		if err != nil {
 			klog.Errorf("%+v", err)
-			if err == importer.ErrRequiresScratchSpace {
-				ds.Close()
-				os.Exit(common.ScratchSpaceNeededExitCode)
-			}
-			err = util.WriteTerminationMessage(fmt.Sprintf("Unable to process data: %+v", err.Error()))
-			if err != nil {
-				klog.Errorf("%+v", err)
-			}
-			ds.Close()
-			os.Exit(1)
 		}
-		touchDoneFile()
-		preallocationApplied = processor.PreallocationApplied()
+
+		return 1
 	}
+	touchDoneFile()
+	// due to the way some data sources can add additional information to termination message
+	// after finished (ds.close() ) termination message has to be written first, before the
+	// the ds is closed
+	// TODO: think about making communication explicit, probably DS interface should be extended
```

Comment:
I dont completly understand if what you did in this commit fixes this issue or not, cause if so I dont see how this changes the behavior than the previous code

Reply:
Let me start from beginning.

I wanted to make some small cleanup the main() so while reading the code one can see the decisions and branches (as it shows there are only two: emptyDisk, and import). 

Then I have exported methods (example: handleImport) and with that exported method I made all local variables really local. That resulted in problems. And that is when I have discovered that we have to first write Import Complete, and the we can close DataSource. 

I think this is designed in a bad way. And I may think about improving that. Some options:
1. we could add a ds.summary() or ds.terminationMessgae(). So then we just do 
```
ds.close() 
additionalInfo:= ds.terminationMessage()
err = importCompleteTerminationMessage(processor.PreallocationApplied(), additionalInfo)
```

2. made ds.close() return termination message?

Anyway, current code works exactly the same way as old. All I managed to do is extract 2 methods (handleempty and handleimport) and make all local variables really local (so this is basically last commit that I needed to drop on https://github.com/kubevirt/containerized-data-importer/pull/1987). And I have discovered that DS and importer are both trying to write to termnation message file. So this is why I have left the comment. So we can see it clearly and do not forget to improve.

Reply:
Thanks for clarifying, getting the termination message after closing sounds good to me.